### PR TITLE
feat(environments): Implement environment filter in stream view

### PR DESCRIPTION
**Pull request by lynnagara**
Also makes "All environments" the default if one is not set, and none of the names matches the list of likely production names

### CODE REVIEW

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

File: src/sentry/static/sentry/app/components/projectHeader/index.jsx [hunk 1]
```diff
@@ -5,63 +5,105 @@ import classNames from 'classnames';
...
   render() {
+    let {project, environments, activeEnvironment} = this.props;
     let navSection = this.props.activeSection;
-    let project = this.props.project;
     let org = this.props.organization;
     let features = new Set(project.features);
     let access = new Set(org.access);
+    let allEnvironmentsLabel = t('All environments');
+
+    // TODO: remove when feature is released
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/components/projectHeader/index.jsx]: add your name to the TODO?

**Review by MaxBittker — APPROVED** [review]
this looks good to me, but I didn't run it locally yet and it's pretty important. could you add some tests to ensure that the change gets propagated and re-runs the search? (might be a hard jest test to write so I can help)

**Comment by getsentry-danger** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">You should update CHANGES due to the size of this PR</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by MaxBittker** [comment]
is this rdy to merge?

**Comment by lynnagara** [comment]
@MaxBittker No, I'm gonna come back to this when we decide what to do with saved searches. Labelled as WIP

File: src/sentry/static/sentry/app/views/stream.jsx [hunk 2]
```diff
@@ -478,6 +523,37 @@ const Stream = createReactClass({
...
+      let environment = context.environment;
+
+      let query = this.state.query;
+
+      if (environment) {
+        query = objToQuery({
+          ...queryToObj(this.state.query),
+          environment: environment.name,
+        });
+      } else {
```
Comment (MaxBittker) [inline src/sentry/static/sentry/app/views/stream.jsx]: again, maybe label these branches explicitly? i know the comment above is doing it, but labeled if branches help a lot when reading dense logic like this IMO

**Review by MaxBittker — APPROVED** [review]
the code itself looks good, and it's really important for dense, important logic like this to be rigorously tested. that being said, i know we had a lot of problems with the tests here I think it's pragmatic to safely merge this behind a feature flag, leave a commit message explaining the change in behavior, do manual testing, and follow up with fixed tests

File: src/sentry/static/sentry/app/views/stream.jsx [hunk 3]
```diff
@@ -345,14 +366,37 @@ const Stream = createReactClass({
 
     let url = this.getGroupListEndpoint();
 
+    let query = this.state.query.replace(/^\s+|\s+$/g, '');
+
+    let activeEnvironment = this.state.activeEnvironment;
+    let activeEnvName = activeEnvironment ? activeEnvironment.name : null;
```
[the remark this answers is not in the document]
Comment (lynnagara) [reply src/sentry/static/sentry/app/views/stream.jsx]: It doesn't make a difference but I thought `null` is a more correct value than `false`

[the remark this answers is not in the document]
Comment (MaxBittker) [reply src/sentry/static/sentry/app/views/stream.jsx]: ok, you're right then

**Pull request merged**